### PR TITLE
feat:merge-lora iterate through bins without loading 

### PR DESCRIPTION
**Pull request by ved1beta**
# Description
feature 1 : The merge-lora script does not load the model into memory, period. It just iterates through each of the bin or safetensors shards and applies the lora to each module as it needs. It's extremely efficient compared to the standard approach.

new file ``` lora_merge_efficient ``` core implementation 
new parameter ``` merge_method ``` : standard /memory efficient

## Motivation and Context

#1679 

## references 
[qlora-pipe/tools/merge_lora.py](https://github.com/tdrussell/qlora-pipe/blob/main/tools/merge_lora.py)

# Tests 
tested with examples/llama-3/qlora-1b.yml 
with tiny llama 1 b instruct and merge_methode:memory efficient

## Release Notes

* **New Features**
  * Added LoRA merge method configuration with a new "memory_efficient" option (default) that processes model shards individually to reduce memory usage, and a "legacy" option that preserves the original full-model merge behavior.
  * New support for simulating NF4 and MoE expert quantization during the LoRA merging process.

### CODE REVIEW

File: src/axolotl/utils/lora_merge_efficient.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    for lora_key, lora_weight in lora_state.items():
+        if clean_key in lora_key:
+            if "lora_A" in lora_key:
+                lora_a = lora_weight
+            elif "lora_B" in lora_key:
+                lora_b = lora_weight
+
+    if lora_a is not None and lora_b is not None:
+        return lora_a, lora_b
+    return None, None
```
[the remark this answers is not in the document]
Comment (winglian) [reply src/axolotl/utils/lora_merge_efficient.py]: Could also use rstrip instead of endswith

File: src/axolotl/utils/lora_merge_efficient.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+    patterns = ["model*.safetensors", "model*.bin", "pytorch_model*.bin"]
+
+    for pattern in patterns:
+        shards.extend(model_path.glob(pattern))
+        if shards:
+            break
+
+    return sorted(shards)
+
```
[the remark this answers is not in the document]
Comment (winglian) [reply src/axolotl/utils/lora_merge_efficient.py:290]: looking in transformers https://github.com/huggingface/transformers/blob/main/src/transformers/utils/__init__.py#L291-L307, it should be sufficient with the `pytorch_model*.bin` and `model*.safetensors`

File: src/axolotl/cli/merge_lora.py [hunk 3]
```diff
@@ -9,18 +9,31 @@
...
+    Merges LoRA adapters with base model using either standard or memory-efficient approach.
 
     Args:
         cfg: Dictionary mapping `axolotl` config keys to values.
     """
+    merge_method = getattr(cfg, "merge_method", "standard")
+    if merge_method == "memory_efficient":
+        _do_merge_lora_efficient(cfg=cfg)
+    else:
+        _do_merge_lora_standard(cfg=cfg)
```
Comment (winglian) [inline src/axolotl/cli/merge_lora.py]: let's be opinionated and prefer the `memory_efficient` merge. the old "`standard`" version can be a manual fallback if users have issues. You could also do something like wrap the `_do_merge_lora_efficient` in a `try/catch` and if it fails, provide a hint to the user to use the `standard` option

Comment (winglian) [reply src/axolotl/cli/merge_lora.py]: maybe rename `standard` to `legacy` or something else since it's no longer axolotl's "standard"

File: src/axolotl/cli/merge_lora.py [hunk 4]
```diff
@@ -9,18 +9,31 @@
...
+
+
+def _do_merge_lora_standard(*, cfg: DictDefault) -> None:
+    """
+    Standard LoRA merging using `merge_and_unload`.
+    Loads the full model into memory before merging.
+    """
+    LOG.info("Using standard LoRA merging method...")
```
Comment (winglian) [inline src/axolotl/cli/merge_lora.py]:
```suggestion
    LOG.debug("Using standard LoRA merging method...")
```

File: src/axolotl/cli/merge_lora.py [hunk 5]
```diff
@@ -49,6 +62,27 @@ def do_merge_lora(*, cfg: DictDefault) -> None:
             processor.save_pretrained(str(Path(cfg.output_dir) / "merged"))
 
 
+def _do_merge_lora_efficient(*, cfg: DictDefault) -> None:
+    """
+    Memory-efficient LoRA merging using shard-by-shard processing.
+    Does not load the full model into memory.
+    """
+    LOG.info("Using memory-efficient LoRA merging method...")
```
Comment (winglian) [inline src/axolotl/cli/merge_lora.py]:
```suggestion
    LOG.debug("Using memory-efficient LoRA merging method...")
```

File: src/axolotl/cli/merge_lora.py [hunk 6]
```diff
@@ -49,6 +62,27 @@ def do_merge_lora(*, cfg: DictDefault) -> None:
...
+    output_path = Path(cfg.output_dir) / "merged"
+    safe_tensors = getattr(cfg, "save_safetensors", True)
+
+    # Perform memory-efficient merge
+    merge_lora_sharded_efficient(
+        base_model_path=cfg.base_model,
+        lora_adapter_path=cfg.lora_model_dir,
+        output_path=output_path,
+        safe_tensors=safe_tensors,
+    )
```
[the remark this answers is not in the document]
Comment (winglian) [reply src/axolotl/cli/merge_lora.py:118]: I don't think this is relevant since `lora_on_cpu` is only when we're loading the full model into memory. Since we're doing it per shard here, it's likely not necessary.

File: src/axolotl/cli/merge_lora.py [hunk 7]
```diff
@@ -49,6 +62,27 @@ def do_merge_lora(*, cfg: DictDefault) -> None:
...
+
+    LOG.info("Memory-efficient LoRA merge completed successfully!")
```
Comment (winglian) [inline src/axolotl/cli/merge_lora.py]:
```suggestion
    LOG.debug("Memory-efficient LoRA merge completed successfully!")
```

File: src/axolotl/cli/merge_lora.py [hunk 8]
```diff
@@ -80,7 +114,7 @@ def do_cli(config: Union[Path, str] = Path("examples/"), **kwargs) -> None:
         parsed_cfg.lora_model_dir = parsed_cfg.output_dir
     if not Path(parsed_cfg.lora_model_dir).exists():
         raise ValueError(
-            f"Target directory for merge: `{parsed_cfg.lora_model_dir}` does not exist."
+            f"Target directory for LoRA merged model does not exist: `{parsed_cfg.lora_model_dir}`"
```
Comment (winglian) [inline src/axolotl/cli/merge_lora.py]:
```suggestion
            f"Target directory for LoRA adapter weights does not exist: `{parsed_cfg.lora_model_dir}`"
```

File: src/axolotl/utils/lora_merge_efficient.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+    os.makedirs(output_path, exist_ok=True)
+
+    config_file = lora_adapter_path / "adapter_config.json"
+    if not config_file.exists():
+        raise FileNotFoundError(f"LoRA config not found: {config_file}")
+
+    lora_config = LoraConfig.from_json_file(config_file)
+    scale = lora_config.lora_alpha / lora_config.r
+
+    LOG.info(f"LoRA scale factor: {scale}")
```
Comment (winglian) [inline src/axolotl/utils/lora_merge_efficient.py]:
```suggestion
    LOG.debug(f"LoRA scale factor: {scale}")
```

File: src/axolotl/utils/lora_merge_efficient.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+
+    lora_file = lora_adapter_path / "adapter_model.safetensors"
+    if not lora_file.exists():
+        lora_file = lora_adapter_path / "adapter_model.bin"
+        if not lora_file.exists():
+            raise FileNotFoundError(
+                f"LoRA adapter weights not found in {lora_adapter_path}"
+            )
+
+    LOG.info(f"Loading LoRA weights from {lora_file}")
```
Comment (winglian) [inline src/axolotl/utils/lora_merge_efficient.py]:
```suggestion
    LOG.debug(f"Loading LoRA weights from {lora_file}")
```

File: src/axolotl/utils/lora_merge_efficient.py [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+    if device != "cpu":
+        LOG.info(f"Moving LoRA weights to {device}")
+        for key, value in tqdm(lora_state.items(), desc="Moving LoRA to device"):
+            lora_state[key] = value.to(device)
+
+    model_shards = get_model_shards(base_model_path)
+    if not model_shards:
+        raise FileNotFoundError(f"No model shards found in {base_model_path}")
+
+    LOG.info(f"Found {len(model_shards)} model shards")
```
Comment (winglian) [inline src/axolotl/utils/lora_merge_efficient.py]:
```suggestion
    LOG.debug(f"Found {len(model_shards)} model shards")
```

File: src/axolotl/utils/lora_merge_efficient.py [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+
+    LOG.info(f"Loading LoRA weights from {lora_file}")
+
+    if lora_file.suffix == ".safetensors":
+        lora_state = safetensors.torch.load_file(lora_file)
+    else:
+        lora_state = torch.load(lora_file, map_location="cpu", weights_only=True)
+
+    if device != "cpu":
+        LOG.info(f"Moving LoRA weights to {device}")
```
Comment (winglian) [inline src/axolotl/utils/lora_merge_efficient.py]:
```suggestion
        LOG.debug(f"Moving LoRA weights to {device}")
```

File: src/axolotl/cli/merge_lora.py (hunk 3 above)
Comment (winglian) [reply src/axolotl/cli/merge_lora.py]: One thing to consider is that this only works for standard LoRA, and not other advanced methods like DoRA

**Comment by djsaunde** [comment]
curious if you have any numbers on how much peak VRAM is saved?

**Comment by ved1beta** [comment]
benchmarks coming soon

**Comment by salmanmohammadi** [comment]
@ved1beta could you also ensure the weights/logits produced by a model which was merged using the legacy vs. memory efficient method are identical?

**Comment by ved1beta** [comment]
this should be insured with the test run ? 
> tested with examples/llama-3/qlora-1b.yml
with tiny llama 1 b instruct and merge_methode:memory efficient

**Comment by winglian** [comment]
Were you able to train a lora, and then merge using both the legacy and memory efficient methods to verify identical merged weights from both methods?

**Comment by ved1beta** [comment]
yes tried merging and everything as you mentioned earlier here is the taining output ```[slacklink]( https://ai-axolotl.slack.com/files/U09BE3G7ZED/F09BNKLDDNZ/untitled?origin_team=T05A3MTMVB8&origin_channel=D09BE3HMM7B )```

i have a claude generated script for testing identical model weights it passes for the given checkpoint generated from the training

**Comment by ved1beta** [comment]
memory usage for both of the merges
calculated with a simple test script 
 ```
Memory-Efficient Method
• Peak GPU Memory: 300 MB
• Peak CPU Memory: 14.4 MB
• Execution Time: 12.0 seconds

Legacy Method
• Peak GPU Memory: 2,914 MB
• Peak CPU Memory: 14.4 MB
• Execution Time: 15.9 seconds
```

**Review by NanoCode012 — APPROVED** [review]

File: src/axolotl/cli/merge_lora.py [hunk 13]
```diff
@@ -4,23 +4,45 @@
...
-    Calls `transformers`' `merge_and_unload` on the model given in the `axolotl` config
-    along with the LoRA adapters to combine them into a single base model.
+    Merges LoRA adapters with base model using either memory-efficient or legacy approach.
 
     Args:
         cfg: Dictionary mapping `axolotl` config keys to values.
     """
+    merge_method = (
+        str(getattr(cfg, "merge_method", "")).strip().lower().replace("-", "_")
+    )
```
Comment (djsaunde) [inline src/axolotl/cli/merge_lora.py]: `merge_method` can only take values: `Literal["legacy", "memory_efficient"]` so you don't need this string handling.

File: src/axolotl/cli/merge_lora.py [hunk 14]
```diff
@@ -4,23 +4,45 @@
...
+    if merge_method in {"legacy", "standard"}:
```
Comment (djsaunde) [inline src/axolotl/cli/merge_lora.py]: "standard" doesn't exist

Comment (djsaunde) [inline src/axolotl/cli/merge_lora.py]:
```suggestion
    if merge_method == "legacy":
```

File: src/axolotl/cli/merge_lora.py [hunk 15]
```diff
@@ -4,23 +4,45 @@
...
+        LOG.debug("Using legacy LoRA merging method...")
+        _do_merge_lora_legacy(cfg=cfg)
+    else:
+        LOG.debug("Using memory-efficient LoRA merging method...")
+        try:
+            _do_merge_lora_efficient(cfg=cfg)
+        except Exception:  # pylint: disable=broad-exception-caught
+            LOG.exception("Memory-efficient merge failed; falling back to legacy.")
+            _do_merge_lora_legacy(cfg=cfg)
```
Comment (djsaunde) [inline src/axolotl/cli/merge_lora.py]: tbh I'd rather have a hard failure here so we know if something is broken

```suggestion
        _do_merge_lora_efficient(cfg=cfg)
```

Comment (djsaunde) [reply src/axolotl/cli/merge_lora.py]: If there are unsupported combinations (you mentioned DoRA, RSLoRA), we should validate this in the pydantic model and raise an error there.

File: src/axolotl/cli/merge_lora.py [hunk 16]
```diff
@@ -52,6 +74,32 @@ def do_merge_lora(*, cfg: DictDefault) -> None:
...
+    Note: Currently only supports standard LoRA, not advanced methods like DoRA or RSLoRA.
+    Will automatically fall back to legacy method for unsupported configurations.
+    """
+    LOG.debug("Using memory-efficient LoRA merging method...")
+
+    output_path = Path(cfg.output_dir) / "merged"
+    safe_tensors = getattr(cfg, "save_safetensors", True)
+    device = "cuda" if torch.cuda.is_available() else "cpu"
+
+    # Perform memory-efficient merge
```
Comment (djsaunde) [inline src/axolotl/cli/merge_lora.py]:
```suggestion
```

File: src/axolotl/utils/schemas/peft.py [hunk 17]
```diff
@@ -140,6 +140,12 @@ class LoraConfig(BaseModel):
     )
 
     merge_lora: bool | None = None
+    merge_method: Literal["legacy", "memory_efficient"] | None = Field(
```
Comment (djsaunde) [inline src/axolotl/utils/schemas/peft.py:158]: slightly prefer `merge_lora_method` so it's unambiguous

File: src/axolotl/cli/utils/lora_merge.py
Comment (djsaunde) [inline src/axolotl/cli/utils/lora_merge.py:1]: I'd suggest moving this under `axolotl/cli/utils/lora_merge.py` along with the legacy method. Then the CLI module can just call either of the two methods based on the config.

File: src/axolotl/utils/lora_merge_efficient.py [hunk 18]
```diff
@@ -0,0 +1,282 @@
...
+
+    LOG.debug(f"Loading LoRA weights from {lora_file}")
+
+    if lora_file.suffix == ".safetensors":
+        lora_state = safetensors.torch.load_file(lora_file)
+    else:
+        try:
+            lora_state = torch.load(lora_file, map_location="cpu", weights_only=True)  # nosec B614
+        except TypeError:
+            lora_state = torch.load(lora_file, map_location="cpu")  # nosec B614
```
Comment (djsaunde) [inline src/axolotl/utils/lora_merge_efficient.py]: why do we need this try/except? can you choose one loading method and stick to it?

File: src/axolotl/utils/lora_merge_efficient.py [hunk 19]
```diff
@@ -0,0 +1,282 @@
...
+    """
+    Memory-efficient LoRA merging that processes shards individually
+    without loading the full model into memory.
+    """
+    base_model_path = Path(base_model_path)
+    lora_adapter_path = Path(lora_adapter_path)
+    output_path = Path(output_path)
+
+    if "/" in str(base_model_path) and not base_model_path.exists():
+        from huggingface_hub import snapshot_download
```
Comment (djsaunde) [inline src/axolotl/utils/lora_merge_efficient.py]: Can be a toplevel import

File: src/axolotl/cli/utils/lora_merge.py [hunk 20]
```diff
@@ -0,0 +1,282 @@
...
+            continue
+        if (
+            filepath.name.startswith("model") and filepath.suffix == ".safetensors"
+        ) or (filepath.name.startswith("pytorch_model") and filepath.suffix == ".bin"):
+            continue
+        if filepath.suffix == ".gguf":
+            continue
+
+        LOG.debug(f"Copying {filepath.name} to output")
+        shutil.copy2(filepath, output_path)
```
Comment (djsaunde) [inline src/axolotl/cli/utils/lora_merge.py:325]: What is the purpose of this? Can you check that we / transformers don't already have a utility for this?

File: src/axolotl/utils/lora_merge_efficient.py [hunk 21]
```diff
@@ -0,0 +1,282 @@
...
+        if shard_path.suffix == ".safetensors":
+            safetensors.torch.save_file(
+                merged_tensors, output_shard_path, metadata=metadata
+            )
+        else:
+            if safe_tensors:
+                LOG.warning(
+                    "safe_tensors=True requested but input shards are .bin; preserving .bin format "
+                    "to avoid index mismatches."
+                )
```
Comment (djsaunde) [inline src/axolotl/utils/lora_merge_efficient.py]: this is a bit confusing. if the user requests safe_tensors, shouldn't we convert them to safetensors?

File: src/axolotl/utils/lora_merge_efficient.py [hunk 22]
```diff
@@ -0,0 +1,282 @@
...
+                        delta = scale * (b_fp32 @ a_fp32)
+                        if bool(
+                            lora_config_dict.get("fan_in_fan_out", False)
+                            or lora_config_dict.get("lora_fan_in_fan_out", False)
+                        ):
+                            delta = delta.T
+                        merged_tensors[key] = (
+                            (base_fp32 + delta).to(original_dtype).detach().cpu()
+                        )
+                        del base_fp32, a_fp32, b_fp32, delta
```
Comment (djsaunde) [inline src/axolotl/utils/lora_merge_efficient.py]: this appears to be duplicated below, can be factored out into a helper method

**Review by djsaunde — CHANGES_REQUESTED** [review]
I saw you added a VRAM / time comparison, can you specify which model you used?

File: src/axolotl/cli/utils/lora_merge.py (hunk 20 above)
Comment (zeusnotfound04) [reply src/axolotl/cli/utils/lora_merge.py:80]: we can remove this if we dont need files like config.json and all in output directory . not sure yes or no

Comment (ved1beta) [reply src/axolotl/cli/utils/lora_merge.py:325]: we can remove this if we dont need files like config.json and all in output directory . not sure yes or no (did'nt found anything similar to this in transformers) for separating non-modules

**Comment by winglian** [comment]
@coderabbit review

**Review by winglian — APPROVED** [review]

**Review by ved1beta — COMMENTED** [review]
LGTM
